File: bindings/kotlin/src/main/kotlin/io/ed3c/aiedge/contracts/v1/InferenceContracts.kt
package io.ed3c.aiedge.contracts.v1

data class ErrorDetail(
    val code: ErrorCode,
    val message: String,
    val retryable: Boolean,
    val provider_id: String? = null,
    val details: Map<String, JsonValue>? = null
) : WireEncodable {
    override fun toWireValue(): Any? {
        val out = linkedMapOf<String, Any?>()
        out["code"] = code.wireValue
        out["message"] = message
        out["retryable"] = retryable
        provider_id?.let { value -> out["provider_id"] = value }
        details?.let { value -> out["details"] = value.mapValues { it.value.toWireValue() } }
        return out
    }
}

data class InferenceEvent(
    val schema: String,
    val request_id: String,
    val sequence: Long,
    val type: InferenceEventType,
    val text_delta: String? = null,
    val tool_proposal: ToolProposal? = null,
    val tool_result: ToolResult? = null,
    val error: ErrorDetail? = null,
    val finish_reason: String? = null
) : WireEncodable {
    override fun toWireValue(): Any? {
        val out = linkedMapOf<String, Any?>()
        out["schema"] = schema
        out["request_id"] = request_id
        out["sequence"] = sequence
        out["type"] = type.wireValue
        text_delta?.let { value -> out["text_delta"] = value }
        tool_proposal?.let { value -> out["tool_proposal"] = value.toWireValue() }
        tool_result?.let { value -> out["tool_result"] = value.toWireValue() }
        error?.let { value -> out["error"] = value.toWireValue() }
        finish_reason?.let { value -> out["finish_reason"] = value }
        return out
    }
}

data class InferenceRequest(
    val schema: String,
    val request_id: String,
    val trace_id: String,
    val task_id: String,
    val task_kind: TaskKind,
    val messages: List<Message>,
    val capability_profile: CapabilityProfile,
    val resource_budget: ResourceBudget,
    val preferred_provider_ids: List<String>? = null,
    val model_ref: ModelArtifactRef? = null
) : WireEncodable {
    override fun toWireValue(): Any? {
        val out = linkedMapOf<String, Any?>()
        out["schema"] = schema
        out["request_id"] = request_id
        out["trace_id"] = trace_id
        out["task_id"] = task_id
        out["task_kind"] = task_kind.wireValue
        out["messages"] = messages.map { item -> item.toWireValue() }
        out["capability_profile"] = capability_profile.toWireValue()
        out["resource_budget"] = resource_budget.toWireValue()
        preferred_provider_ids?.let { value -> out["preferred_provider_ids"] = value.map { item -> item } }
        model_ref?.let { value -> out["model_ref"] = value.toWireValue() }
        return out
    }
}

data class ToolResult(
    val proposal_id: String,
    val state: ResultState,
    val output: JsonValue? = null,
    val error: ErrorDetail? = null,
    val receipt_subject: SubjectRef? = null
) : WireEncodable {
    override fun toWireValue(): Any? {
        val out = linkedMapOf<String, Any?>()
        out["proposal_id"] = proposal_id
        out["state"] = state.wireValue
        output?.let { value -> out["output"] = value.toWireValue() }
        error?.let { value -> out["error"] = value.toWireValue() }
        receipt_subject?.let { value -> out["receipt_subject"] = value.toWireValue() }
        return out
    }
}

File: bindings/swift/Sources/AIEdgeContracts/InferenceContracts.swift
import Foundation

/// Typed error shared across provider boundaries.
public struct ErrorDetail: Codable, Equatable, Sendable {
    public let code: ErrorCode
    public let message: String
    public let retryable: Bool
    public let provider_id: String?
    public let details: [String: JSONValue]?

    public init(
        code: ErrorCode,
        message: String,
        retryable: Bool,
        provider_id: String? = nil,
        details: [String: JSONValue]? = nil
    ) {
        self.code = code
        self.message = message
        self.retryable = retryable
        self.provider_id = provider_id
        self.details = details
    }
}

/// Ordered streaming event envelope.
public struct InferenceEvent: Codable, Equatable, Sendable {
    public let schema: String
    public let request_id: String
    public let sequence: Int64
    public let type: InferenceEventType
    public let text_delta: String?
    public let tool_proposal: ToolProposal?
    public let tool_result: ToolResult?
    public let error: ErrorDetail?
    public let finish_reason: String?

    public init(
        schema: String,
        request_id: String,
        sequence: Int64,
        type: InferenceEventType,
        text_delta: String? = nil,
        tool_proposal: ToolProposal? = nil,
        tool_result: ToolResult? = nil,
        error: ErrorDetail? = nil,
        finish_reason: String? = nil
    ) {
        self.schema = schema
        self.request_id = request_id
        self.sequence = sequence
        self.type = type
        self.text_delta = text_delta
        self.tool_proposal = tool_proposal
        self.tool_result = tool_result
        self.error = error
        self.finish_reason = finish_reason
    }
}

/// Top-level inference request. Host policy evaluates it before provider selection.
public struct InferenceRequest: Codable, Equatable, Sendable {
    public let schema: String
    public let request_id: String
    public let trace_id: String
    public let task_id: String
    public let task_kind: TaskKind
    public let messages: [Message]
    public let capability_profile: CapabilityProfile
    public let resource_budget: ResourceBudget
    public let preferred_provider_ids: [String]?
    public let model_ref: ModelArtifactRef?

    public init(
        schema: String,
        request_id: String,
        trace_id: String,
        task_id: String,
        task_kind: TaskKind,
        messages: [Message],
        capability_profile: CapabilityProfile,
        resource_budget: ResourceBudget,
        preferred_provider_ids: [String]? = nil,
        model_ref: ModelArtifactRef? = nil
    ) {
        self.schema = schema
        self.request_id = request_id
        self.trace_id = trace_id
        self.task_id = task_id
        self.task_kind = task_kind
        self.messages = messages
        self.capability_profile = capability_profile
        self.resource_budget = resource_budget
        self.preferred_provider_ids = preferred_provider_ids
        self.model_ref = model_ref
    }
}

/// Validated tool outcome.
public struct ToolResult: Codable, Equatable, Sendable {
    public let proposal_id: String
    public let state: ResultState
    public let output: JSONValue?
    public let error: ErrorDetail?
    public let receipt_subject: SubjectRef?

    public init(
        proposal_id: String,
        state: ResultState,
        output: JSONValue? = nil,
        error: ErrorDetail? = nil,
        receipt_subject: SubjectRef? = nil
    ) {
        self.proposal_id = proposal_id
        self.state = state
        self.output = output
        self.error = error
        self.receipt_subject = receipt_subject
    }
}
